### PR TITLE
#2550 rollover test fixes

### PR DESCRIPTION
**Pull request by stenington**
I think this will fix some tests that previously relied on config in the .env matching the actual date, i.e the `DATES_*_YEAR` config needed to agree with `Season.year.current` or tests would fail.

It also fixes some tests that started to break when I set the season rollover date to be June 1. Doing so somehow changed the division calculations so that students intended to be in the junior division were being calculated as senior.

### CODE REVIEW

**Comment by stenington** [comment]
@shaunxp20 I'm happy for you to review this one for readability without worrying about changing your config to simulate the season rollover. If you want to check the rollover, what I did was leave all my date config in the `.env` as 2020, but make `DATES_NEW_SEASON_START_MONTH` and `DATES_NEW_SEASON_START_DAY` something in the past, 7/4 in my case. That way the software generally thinks it's the 2021 season, except where this PR makes changes to force it to think it's something else.

File: spec/models/account_spec.rb [hunk 1]
```diff
@@ -2,6 +2,11 @@
 require "./lib/fill_pdfs"
 
 RSpec.describe Account do
+  let(:season_with_templates) { Season.new(2020) }
+  before {
+    allow(Season).to receive(:current).and_return(season_with_templates)
```
Comment (shaunxp20) [inline spec/models/account_spec.rb:7]: 🤔 I see this pattern in various places. Just curious, wondering why the real `Season` is being called, and then stubbed? I'm probably missing something 😊 why not use the real `Season.current`? Or not use the real `Season` at all and return a double instead? 🤔 😸 Also, quibble at the block syntax, in Ruby the convention is multi-line block are do/end, and single-line blocks use the curlys, fwiw. 😁

File: spec/models/team_spec.rb [hunk 2]
```diff
@@ -234,7 +234,7 @@
     team.reload
 
     old_student = FactoryBot.create(:student, date_of_birth: 15.years.ago)
-    young_student = FactoryBot.create(:student, date_of_birth: 14.years.ago)
+    young_student = FactoryBot.create(:student, date_of_birth: 13.years.ago)
```
Comment (shaunxp20) [inline spec/models/team_spec.rb:237]: These age changes also caught my eye. 👀 Just curious why they had to change? 🤔 😄

File: spec/models/account_spec.rb (hunk 1 above)
Comment (stenington) [reply spec/models/account_spec.rb:7]: > why not use the real Season.current

That's because e.g. this year when the season ticks over to 2021, there will be no 2021 certificate templates in the repo yet, and the test suite will fail until they're added, which isn't particularly helpful in my opinion. Especially because we won't have real 2021 certificates at that point, so we'll just copy 2020 certificates as 2021 anyway most likely.

>Or not use the real Season at all and return a double instead

Yeah, could do that. Would there be an advantage to using a double here?

> quibble at the block syntax

Oh cool, I will try to follow that convention. I wasn't aware of it.

File: spec/models/team_spec.rb (hunk 2 above)
Comment (stenington) [reply spec/models/team_spec.rb:237]: 14 was too close to the boundary condition I guess. When I changed the season rollover date to early July, it broke the tests I expected it to break, but it also made the platform think `young_student`, in this case, was a senior division student when they need to be a junior division student. So making them even younger fixed that.

File: spec/models/account_spec.rb (hunk 1 above)
Comment (shaunxp20) [reply spec/models/account_spec.rb:7]: > > why not use the real Season.current
> 
> That's because e.g. this year when the season ticks over to 2021, there will be no 2021 certificate templates in the repo yet, and the test suite will fail until they're added, which isn't particularly helpful in my opinion. Especially because we won't have real 2021 certificates at that point, so we'll just copy 2020 certificates as 2021 anyway most likely.

My "why not use the real Season current" comment was referring to the line two lines above where we're setting the season via `Season.new`; I think I'm missing something, what's the point of stubbing `Season.current` to the thing that was just set?

> > Or not use the real Season at all and return a double instead
> 
> Yeah, could do that. Would there be an advantage to using a double here?

Imho, the advantage would clarity; and more isolation from `Season`, or fully committing to `Season`. Basically, my initial inquiry was just that I'm curious/confused why the mix of real vs stub, why not one or the other? 🤔

File: spec/models/team_spec.rb (hunk 2 above)
Comment (shaunxp20) [reply spec/models/team_spec.rb:237]: Huh, that seems odd! 👃 😂 👍 👌

File: spec/models/account_spec.rb (hunk 1 above)
Comment (stenington) [reply spec/models/account_spec.rb:7]: Oh, because later in the year `Season.current` will return 2021 which won't have templates and will break. Specifying 2020 is redundant for now, but will allow these tests to continue to pass in the future.

I'll try putting in a double instead and seeing if that makes more sense. I don't have a principled reason for using the real Season, it's just what came out of my investigating and allowed these tests to pass even when I configured the app to think the 2021 season had started.

File: spec/models/team_spec.rb (hunk 2 above)
Comment (stenington) [reply spec/models/team_spec.rb:237]: Yeah, it's very odd! Actually it makes me nervous that there's some weird boundary condition bug lurking in the app around divisions, I was considering opening a ticket for further exploration of that maybe.
